### PR TITLE
Devas1768 reply side comment

### PR DESCRIPTION
**Pull request by cedrtang**
Add function to reply to a comment done by harvesting

### CODE REVIEW

File: assembl/static2/css/components/fictionSideCommentBox.scss [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+      }
+    }
+
+    .submit-comment {
+      padding: 0 15px;
+    }
+  }
+
+  .arrow-left {
+    position: absolute;
```
Comment (MilanBarande) [inline assembl/static2/css/components/fictionSideCommentBox.scss]: I see quite a few ```position: absolute;``` here, is this all responsive 😛 ?

File: assembl/static2/js/app/components/debate/brightMirror/sideComment/innerBoxSubmit.jsx [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+import AvatarImage from '../../../common/avatarImage';
+import RichTextEditor from '../../../common/richTextEditor';
+import { FICTION_COMMENT_MAX_LENGTH } from '../../../../constants';
+
+export type Props = {
+  userId: ?string,
+  userName: ?string,
+  body: EditorState,
+  updateBody: Function,
```
Comment (MilanBarande) [inline assembl/static2/js/app/components/debate/brightMirror/sideComment/innerBoxSubmit.jsx]: Maybe describe this function's shape

File: assembl/static2/js/app/components/debate/brightMirror/sideComment/innerBoxView.jsx [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+import AvatarImage from '../../../common/avatarImage';
+import { transformLinksInHtml /* getUrls */ } from '../../../../utils/linkify';
+import { postBodyReplacementComponents } from '../../common/post/postBody';
+
+export type Props = {
+  contentLocale: string,
+  extractIndex: number,
+  extracts: ?Array<FictionExtractFragment>,
+  comment: ExtractComment,
+  changeCurrentExtract: ?Function
```
Comment (MilanBarande) [inline assembl/static2/js/app/components/debate/brightMirror/sideComment/innerBoxView.jsx]: same as the other, Functions which arent't 'refetch' and things like that should be described ideally :)

Comment (amensouissi) [inline assembl/static2/js/app/components/debate/brightMirror/sideComment/innerBoxView.jsx]: (number)=>void

File: assembl/static2/js/app/components/debate/brightMirror/sideComment/sideCommentBox.jsx [hunk 4]
```diff
@@ -147,6 +147,16 @@ class DumbSideCommentBox extends React.Component<Props, State> {
     return extracts && extracts.length > 0 ? extracts[extractIndex] : null;
   };
 
+  getCurrentComment = (extract) => {
```
Comment (amensouissi) [inline assembl/static2/js/app/components/debate/brightMirror/sideComment/sideCommentBox.jsx]: flow?

File: assembl/static2/js/app/components/debate/brightMirror/sideComment/sideCommentBox.jsx [hunk 5]
```diff
@@ -147,6 +147,16 @@ class DumbSideCommentBox extends React.Component<Props, State> {
...
+    const topComments = extract && extract.comments && extract.comments.filter(post => post.parentId === null);
+    return topComments && topComments.length > 0 ? topComments[0] : null;
+  };
+
+  getCurrentCommentReply = (extract, comment) => {
```
Comment (amensouissi) [inline assembl/static2/js/app/components/debate/brightMirror/sideComment/sideCommentBox.jsx]: and here :)

File: assembl/static2/js/app/components/debate/brightMirror/sideComment/sideCommentBox.jsx [hunk 6]
```diff
@@ -147,6 +147,16 @@ class DumbSideCommentBox extends React.Component<Props, State> {
     return extracts && extracts.length > 0 ? extracts[extractIndex] : null;
   };
 
+  getCurrentComment = (extract) => {
+    const topComments = extract && extract.comments && extract.comments.filter(post => post.parentId === null);
+    return topComments && topComments.length > 0 ? topComments[0] : null;
```
Comment (amensouissi) [inline assembl/static2/js/app/components/debate/brightMirror/sideComment/sideCommentBox.jsx]: (topComments && topComments[0])  || null

File: assembl/static2/js/app/components/debate/brightMirror/sideComment/sideCommentBox.jsx [hunk 7]
```diff
@@ -147,6 +147,16 @@ class DumbSideCommentBox extends React.Component<Props, State> {
...
+  };
+
+  getCurrentCommentReply = (extract, comment) => {
+    const replies = extract && comment && extract.comments.filter(post => post.parentId === comment.id);
+    return replies && replies.length > 0 ? replies[0] : null;
```
Comment (amensouissi) [inline assembl/static2/js/app/components/debate/brightMirror/sideComment/sideCommentBox.jsx]: (replies && replies[0]) || null

File: assembl/static2/js/app/components/debate/brightMirror/sideComment/sideCommentBox.jsx [hunk 8]
```diff
@@ -195,6 +205,35 @@ class DumbSideCommentBox extends React.Component<Props, State> {
...
+      extractId: currentExtract.id,
+      parentId: currentComment.id,
+      body: convertContentStateToHTML(body.getCurrentContent()),
+      contentLocale: contentLocale
+    };
+    createPost({ variables: postVars }).then(() => {
+      this.setState({
+        replying: false
+      });
+      window.getSelection().removeAllRanges();
```
Comment (amensouissi) [inline assembl/static2/js/app/components/debate/brightMirror/sideComment/sideCommentBox.jsx]:
```
this.setState({
         replying: false
       }, () => {
         window.getSelection().removeAllRanges();
         displayAlert('success', I18n.t('debate.brightMirror.sideCommentSuccessMsg'));
         refetchPost();
});
```

File: assembl/static2/js/app/components/debate/brightMirror/sideComment/innerBoxView.jsx [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+};
+
+const renderRichtext = text => activeHtml(text && transformLinksInHtml(text), postBodyReplacementComponents());
+
+const InnerBoxView = ({ contentLocale, extractIndex, extracts, comment, changeCurrentExtract }: Props) => {
+  const displayName =
+    comment && comment.creator && !comment.creator.isDeleted ? comment.creator.displayName : I18n.t('deletedUser');
+
+  return (
+    <div>
```
Comment (MilanBarande) [inline assembl/static2/js/app/components/debate/brightMirror/sideComment/innerBoxView.jsx]: Fragment?

File: assembl/static2/js/app/components/debate/brightMirror/sideComment/sideCommentBox.jsx (hunk 4 above)
Comment (MilanBarande) [inline assembl/static2/js/app/components/debate/brightMirror/sideComment/sideCommentBox.jsx]: ```extract: string``` ?

File: assembl/static2/js/app/components/debate/brightMirror/sideComment/sideCommentBox.jsx (hunk 5 above)
Comment (MilanBarande) [inline assembl/static2/js/app/components/debate/brightMirror/sideComment/sideCommentBox.jsx]: same, quick annotations:)

File: assembl/static2/js/app/components/debate/brightMirror/sideComment/sideCommentBox.jsx (hunk 8 above)
Comment (MilanBarande) [inline assembl/static2/js/app/components/debate/brightMirror/sideComment/sideCommentBox.jsx]: maybe put whatever comes after this ```this.setState```in the setState callback function ? I think it's better. @amensouissi can you confirm?

Comment (MilanBarande) [reply assembl/static2/js/app/components/debate/brightMirror/sideComment/sideCommentBox.jsx]: I believe it's about the timing, if you need those to run only once the state is updated

File: assembl/static2/js/app/components/debate/brightMirror/sideComment/sideCommentBox.jsx [hunk 10]
```diff
@@ -255,83 +307,49 @@ class DumbSideCommentBox extends React.Component<Props, State> {
...
-                  maxLength={FICTION_COMMENT_MAX_LENGTH}
-                  onChange={this.updateBody}
-                  placeholder={I18n.t('debate.brightMirror.commentLabel')}
-                  withAttachmentButton
-                />
-              </div>
-            )}
-          </div>
+          )}
+          {submitting ? (
```
Comment (amensouissi) [inline assembl/static2/js/app/components/debate/brightMirror/sideComment/sideCommentBox.jsx]: add a new component ExtractCommentForm ?

File: assembl/static2/js/app/components/debate/brightMirror/sideComment/sideCommentBox.jsx [hunk 11]
```diff
@@ -224,18 +263,31 @@ class DumbSideCommentBox extends React.Component<Props, State> {
...
+  toggleReplying = () => {
+    this.setState(state => ({ replying: !state.replying }));
+  };
+
+  renderActionFooter = () => {
+    const { userCanReply } = this.props;
+    const { extractIndex } = this.state;
+    const currentExtract = this.getCurrentExtract(extractIndex);
+    const currentComment = this.getCurrentComment(currentExtract);
+    const hasReply = !!this.getCurrentCommentReply(currentExtract, currentComment);
```
Comment (MilanBarande) [inline assembl/static2/js/app/components/debate/brightMirror/sideComment/sideCommentBox.jsx]: Curious: why ```!!``` ?

Comment (ayyazdaniaryan) [reply assembl/static2/js/app/components/debate/brightMirror/sideComment/sideCommentBox.jsx]: It's JS etiquette for turning a non-bool into a bool.

File: assembl/static2/js/app/components/debate/common/replyToCommentButton.jsx [hunk 12]
```diff
@@ -1,16 +1,22 @@
...
   /** On click callback function */
-  onClickCallback: Function
+  onClickCallback: Function,
+  disabled?: boolean
 };
 
-const ReplyToCommentButton = ({ onClickCallback }: ReplyToCommentButtonProps) => (
-  <Link className="action-reply" onClick={onClickCallback}>
+const ReplyToCommentButton = ({ onClickCallback, disabled }: ReplyToCommentButtonProps) => (
+  <Link className={classNames('action-reply', { disabled: disabled })} onClick={!disabled ? onClickCallback : null}>
```
Comment (MilanBarande) [inline assembl/static2/js/app/components/debate/common/replyToCommentButton.jsx]: about the onClick: isn't the css disabling of the button enough ?

File: assembl/tests/views_tests/test_graphql_posts.py [hunk 13]
```diff
@@ -34,4 +34,38 @@ def test_mutation_add_extract_comment(admin_user, graphql_request, idea_in_threa
...
+    from assembl.models import Post, ExtractComment
+    post = Post.get(raw_id)
+    post.extracts.append(extract_post_1_to_subidea_1_1)
+    post.db.flush()
+
+    extract_id = extract_post_1_to_subidea_1_1.graphene_id()
+    comment_id = extract_comment.graphene_id()
+    idea_id = idea_in_thread_phase
+    res = schema.execute(u"""
+        mutation myFirstMutation {
```
Comment (amensouissi) [inline assembl/tests/views_tests/test_graphql_posts.py]: myFirstMutation ?

File: assembl/static2/js/app/stories/components/debate/brightMirror/sideComment/sideCommentBox.stories.jsx [hunk 14]
```diff
@@ -145,22 +193,36 @@ export const defaultSideCommentBox: SideCommentBoxProps = {
   toggleExtractsBox: Function,
   position: { x: 0, y: 0 },
   setPositionToExtract: Function,
-  clearHighlights: Function
+  clearHighlights: Function,
```
Comment (amensouissi) [inline assembl/static2/js/app/stories/components/debate/brightMirror/sideComment/sideCommentBox.stories.jsx]: Function ?! confusion

File: assembl/static2/js/app/components/debate/brightMirror/sideComment/sideCommentBox.jsx (hunk 11 above)
Comment (MilanBarande) [reply assembl/static2/js/app/components/debate/brightMirror/sideComment/sideCommentBox.jsx]: oooh right

File: assembl/static2/js/app/components/debate/brightMirror/sideComment/sideCommentBox.jsx [hunk 15]
```diff
@@ -39,13 +36,15 @@ export type Props = {
   toggleCommentsBox?: Function,
   clearHighlights: Function,
   position: { x: number, y: number },
-  setPositionToExtract: Function
+  setPositionToExtract: Function,
```
Comment (pascalhuynh) [inline assembl/static2/js/app/components/debate/brightMirror/sideComment/sideCommentBox.jsx]: add Function description (if possible)

File: assembl/static2/js/app/components/debate/brightMirror/sideComment/sideCommentMenu.jsx [hunk 16]
```diff
@@ -18,7 +18,8 @@ type Props = {
   handleClickAnchor: Function,
   cancelSubmit: Function,
   refetchPost: Function,
-  setCommentBadgeToExtractPosition: Function
+  setCommentBadgeToExtractPosition: Function,
```
Comment (pascalhuynh) [inline assembl/static2/js/app/components/debate/brightMirror/sideComment/sideCommentMenu.jsx]: add Function description (if possible)

File: assembl/static2/js/app/stories/components/debate/brightMirror/sideComment/innerBoxSubmit.stories.jsx [hunk 17]
```diff
@@ -0,0 +1,30 @@
+// @flow
+import React from 'react';
+/* eslint-disable import/no-extraneous-dependencies */
+import { storiesOf } from '@storybook/react';
```
Comment (pascalhuynh) [inline assembl/static2/js/app/stories/components/debate/brightMirror/sideComment/innerBoxSubmit.stories.jsx]: add `import { action } from '@storybook/addon-actions';` below

File: assembl/static2/js/app/stories/components/debate/brightMirror/sideComment/innerBoxSubmit.stories.jsx [hunk 18]
```diff
@@ -0,0 +1,30 @@
...
+/* eslint-enable */
+
+import InnerBoxSubmit, { Props } from '../../../../../components/debate/brightMirror/sideComment/innerBoxSubmit';
+
+export const defaultInnerBoxSubmitProps: Props = {
+  userId: '1',
+  userName: 'Odile Deray',
+  body: EditorState.createWithContent(ContentState.createFromText('Hello')),
+  updateBody: () => {},
+  submit: Function,
```
[the remark this answers is not in the document]
Comment (pascalhuynh) [reply assembl/static2/js/app/stories/components/debate/brightMirror/sideComment/innerBoxSubmit.stories.jsx]: change `submit: Function` to `submit: action('submit')`

File: assembl/static2/js/app/stories/components/debate/brightMirror/sideComment/innerBoxSubmit.stories.jsx [hunk 19]
```diff
@@ -0,0 +1,30 @@
...
+  cancelSubmit: Function
```
Comment (pascalhuynh) [inline assembl/static2/js/app/stories/components/debate/brightMirror/sideComment/innerBoxSubmit.stories.jsx]: change `cancelSubmit: Function` to `cancelSubmit: action('cancelSubmit')`

File: assembl/static2/js/app/stories/components/debate/brightMirror/sideComment/sideCommentBox.stories.jsx (hunk 14 above)
Comment (pascalhuynh) [reply assembl/static2/js/app/stories/components/debate/brightMirror/sideComment/sideCommentBox.stories.jsx]: change Function to storybook action (like above)

File: assembl/static2/css/components/fictionSideCommentBox.scss (hunk 1 above)
Comment (cedrtang) [reply assembl/static2/css/components/fictionSideCommentBox.scss]: Nope, its a feature only for desktop

File: assembl/static2/js/app/components/debate/common/replyToCommentButton.jsx (hunk 12 above)
Comment (cedrtang) [reply assembl/static2/js/app/components/debate/common/replyToCommentButton.jsx]: @MilanBarande @pascalhuynh If it was a button, I could use the 'disabled' selector but here it's a link so I have to cancel the callback and add a css class to change color

Comment (cedrtang) [reply assembl/static2/js/app/components/debate/common/replyToCommentButton.jsx]: It s a bootstrap feature that doesn't work with links like this, without href but this works in css: pointer-events: none;
  cursor: default;

File: assembl/static2/js/app/components/debate/brightMirror/sideComment/sideCommentBox.jsx (hunk 10 above)
Comment (cedrtang) [reply assembl/static2/js/app/components/debate/brightMirror/sideComment/sideCommentBox.jsx]: InnerBoxSubmit has been created for this, to be used for the comments and the replies

**Comment by cedrtang** [comment]
@amensouissi Just to let you know, I found a bug where a side comment which is linked to an extract was shown as a fiction. If it has a parentId it then is shown as a comment in debate section. I think we're reusing so much the same tables and types that it s all intertwined and create bugs in many different places that are hard to track.

cc: @pascalhuynh

**Pull request merged**